### PR TITLE
Add Validation and Sort function to ZooQueueLock

### PR DESCRIPTION
**Pull request by Manno15**
A possible solution to #1935. This alters `ZooQueueLock` prefix to match the expected prefix in `ZooLock.validateAndSortChildrenByLockPrefix`. This does get rid of the warnings found in #1935 and so far everything seems to be working as expected. 

I am marking it as a draft since I want a better mechanism for the UUID creation but I haven't run into any issues from some basic testing. This PR can also be the starting point for further improvements that were listed in #1935.

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooQueueLock.java
Comment (EdColeman) [inline core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooQueueLock.java:36]: If it is required that this PREFIX here and the ZLOCK_PREFIX in ZooLock match, then maybe the definition should be placed in Constants and shared.  (Not saying that the should be the same, that's a different discussion)

Comment (Manno15) [reply core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooQueueLock.java:36]: I think the parts that have to match is the "zlock#" and then also end in a "#". The first part could definitely be placed in Constants. 

I guess the issue is the `CreateTable` command uses `ZooQueueLock` for creation and the `FateCommand` uses `ZooLock` for validation.

**Review by EdColeman — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooQueueLock.java
Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooQueueLock.java:36]: Constants.java has slowly been in the process of being phased out, as constants move closer to the code that is responsible for them, under the premise that fewer global constants increase modularity and separation of responsibilities.

If the schema is defined by ZooLock and ZooQueueLock is trying to match it, then the constant should live in ZooLock.

**Comment by Manno15** [comment]
After doing more investigating, I am leaning towards the side of keeping the use of `ZooQueueLock` as is and not trying to refactor it to use `ZooLock`. Instead, I'd rather alter the validation and sorting process of the locks. In `AdminUtil` (which is called from the Fate Print command) we call `ZooLock.validateAndSortChildrenByLockPrefix`. This expects the lock prefix to only be zLock which is where the warning is coming from. If the goal of this function is just to validate and sort the locks then I think broadening its use case by accepting different lock prefixes can solve this issue while not forcing every lock to look like ZooLocks. I would do this by moving out this function into a utility class or in its own class. Then, whichever class needs to validate/sort locks can call it and either pass in the expected lock prefix or pull from a list of valid ones. This is under the assumption that we want to keep other lock classes like `ZooQueueLock`/`DistributedReadWriteLock` (which can have different prefixes) and not convert everything to `ZooLock`.

**Comment by ctubbsii** [comment]
If the offending code is the validation in AdminUtil, then it probably got introduced in my fix for #1926. I probably failed to realize that the lock node used for Fate locks was a different path and syntax than other locks, because all the lock node paths are passed around as Strings, instead a more specific type.

It might be beneficial to change the code around locking to have more specific types, to prevent this in future. I'm thinking something like:

```java
FateLockPath fLockPath = FateLock.path(somePathString);
ZooQueueLock.validateAndSort(fLockPath, children);

ZooLockPath zLockPath = ZooLock.path(somePathString);
ZooLock.validateAndSort(zLockPath, children);
```

Doing something like this would ensure we can tell the difference between a lock path used for ZooQueueLock and one used for ZooLock, which are very different.

**Comment by Manno15** [comment]
Yeah, I was thinking along those lines. I figured having one function that could handle multiple cases would be easier to manage in the long run but I see the benefit of each having its own validation function.

**Comment by dlmarion** [comment]
`ZooLock.validateAndSortChildrenByLockPrefix` validates and sorts the locks at the associated path for ZooLock's purposes. If the ephemeral node name does not meet the format specifications of ZooLock, then maybe we should just make that a `trace` level event instead of `warn`. The list of valid children is printed at the `trace` level at the end of the method, maybe the invalid children should be the same. My suggestion applies to the log message [here](https://github.com/apache/accumulo/blob/main/core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooLock.java#L215) and [here](https://github.com/apache/accumulo/blob/main/core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooLock.java#L218)

**Comment by ctubbsii** [comment]
> `ZooLock.validateAndSortChildrenByLockPrefix` validates and sorts the locks at the associated path for ZooLock's purposes.

Right. The problem here is that there was a coding error that incorrectly interpreted this lock path as being one where ZooLock's locks are stored... but that shouldn't have been the case. I don't think we should merely log... we should actively prevent such coding errors from occurring by avoiding String-based parameters in ZooLock (and ZooQueueLock) APIs, where possible.

**Comment by Manno15** [comment]
In my latest commits, I added a validation and sort method inside `ZooQueueLock`. It pretty much the same as the one in `ZooLock` so it is possible we might want to simplify it some more, but it works as-is. There was only the call in `AdminUtil` that I needed to change to use the new validation function. I also labeled the various areas where we call any of the validation functions with what kind of lock type it is using (Just the ones that are outside of `ZooLock` and its direct tests).

**Comment by Manno15** [comment]
@ctubbsii @EdColeman Thoughts on my most recent changes? This should be ready for review again.

**Comment by Manno15** [comment]
Do you know of a case for when that would occur? I am trying to think of a way to test it but I am not sure. From what I can tell, there are only a few areas of the code where we use `ZooQueueLock` (things like CreateTable use it) while `ZooLock` is used the majority of the time.

**Comment by dlmarion** [comment]
Wasn't that the issue in https://github.com/apache/accumulo/issues/1935 ?

**Comment by Manno15** [comment]
Most of the issue was using the wrong verification function inside `AdminUtil`. It called the ZooLock one whereas it only dealt with Fate locks which use `ZooQueueLock`.

**Comment by dlmarion** [comment]
Ok, I see, I retract my statement then.

File: core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooQueueLock.java
Comment (EdColeman) [inline core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooQueueLock.java:137]: It may be more descriptive if, instead of Child... if it was FateLock (or flock) or something that is more descriptive than generic child - this applies in multiple places.

**Review by EdColeman — COMMENTED** [review]
Other than a suggestion for minor rewording of log statements to be more descriptive, this looks okay.

File: core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooQueueLock.java
Comment (Manno15) [reply core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooQueueLock.java]: I made some log changes in my latest commit

Comment (Manno15) [reply core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooQueueLock.java]: @EdColeman Thoughts on my recent log changes? Is that more descriptive and do you think I should make those same log changes in the validation function inside `ZooLock`? If you approve then I will get this merged in today.

**Review by EdColeman — COMMENTED** [review]
Thanks for the changes.

File: core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooQueueLock.java
Comment (EdColeman) [reply core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooQueueLock.java]: I'll look at zoo lock.  That should not hold up this request.  I think this can be merged unless others have issues.

**Review by EdColeman — APPROVED** [review]

**Pull request merged**
